tests/unit/id.cpp

```diff
@@ -215,6 +215,17 @@ void grb_program2( const struct input &in, struct output &out ) {
 		return;
 	}
 
+	/*
+	Test for move assignement id cleanup.
+	
+	Creating and performing move assignment on multiple new objects to generate potential id collision.
+	*/
+	for( int i = 0; i < 1000; i++ ) {
+		grb::Matrix< int > new_one( 10, 10, 10 );
+		grb::Matrix< int > new_two( 10, 10, 10 );
+		new_two = std::move(new_one);
```

Comment:
I'm confused on how a clash is detected -- does it on collision trigger some internal assertion? If yes, is something like the below a correct way to make the test more robust?
```
	for( int i = 0; i < 1000; i++ ) {
		grb::Matrix< int > new_one( 10, 10, 10 );
		grb::Matrix< int > new_two( 10, 10, 10 );
                if( grb::getID( new_one ) == grb::getID( new_two ) { /* test failed */ }
                const size_t old_one_id = grb::getID( new_one );
                const size_t old_two_id = grb::getID( new_two );
		new_two = std::move(new_one);
                if( grb::getID( new_two ) == old_two_id ) { /* test fails */ }
                if( grb::getID( new_two ) != old_one_id ) { /* test fails */ }
```
(Note that I don't think it's correct to call getID(new_one) after it has been moved from, that would incur undefined behaviour. I also don't think it is correct to build an `std::set` of previous IDs, because in principle our spec allows re-use of IDs, right?)

Reply:
I think that the issue here is that when a matrix is created, in the `initialize` function, the pointer `alloc[0]` is forwarded to the `reference_mapper`:

https://github.com/Algebraic-Programming/ALP/blob/dc0396cbbae05aaa4df8d90dafe50c938166bfa9/include/graphblas/reference/matrix.hpp#L1426-L1434

And then, the following assertion of the `reference_mapper` fails:

https://github.com/Algebraic-Programming/ALP/blob/dc0396cbbae05aaa4df8d90dafe50c938166bfa9/include/graphblas/utils/DMapper.hpp#L91-L98

I think that the problem is that, upon move assignment, the `old_two_id` isn't removed from the `reference_mapper`, which causes old unused pointers to remain stored in the map, in turn causing the assertion to fail on one of the subsequent matrix creations.

Reply:
Yeah I agree but I mean the that a unit test should not rely on assertions -- the error should ideally also cause an observable ill effect that the test can check for, such as those in the above suggested code block

Reply:
I'm not certain if the code after the failed assert causes any ill effect that is observable outside of the `DMapper.h` itself. 

The maps will have some outdated information, i.e. two deterministic IDs pointing to the same non-deterministic ID, but I don't think this could cause a problem, as the outdated deterministic ID shouldn't be retrievable from any `getID` call.

Reply:
> I'm not certain if the code after the failed assert causes any ill effect that is observable outside of the DMapper.h itself.

I agree -- I meant a unit test should not rely on the implementation (here asserts) to detect bugs. I have resolved this one as suggested above: it will do a bunch of moves and checks that those moves are consistent in terms of observable behaviour. Running the unit tests with that change now and if it works, will merge. Many thanks for this one (and apologies on the big delay in getting these MRs done!)